plugin/storage/es/spanstore/reader.go

```diff
@@ -205,39 +205,76 @@ func (s *SpanReader) FindTraces(traceQuery *spanstore.TraceQueryParameters) ([]*
 }
 
 func (s *SpanReader) multiRead(traceIDs []string, startTime, endTime time.Time) ([]*model.Trace, error) {
+
 	if len(traceIDs) == 0 {
 		return []*model.Trace{}, nil
 	}
 	searchRequests := make([]*elastic.SearchRequest, len(traceIDs))
 
-	for i, traceID := range traceIDs {
-		query := elastic.NewTermQuery("traceID", traceID)
-		searchRequests[i] = elastic.NewSearchRequest().IgnoreUnavailable(true).Type("span").
-			Source(elastic.NewSearchSource().Query(query).Size(defaultDocCount))
-	}
-
+	var traces []*model.Trace
 	// Add an hour in both directions so that traces that straddle two indexes are retrieved.
-	// ie starts in one and ends in another.
+	// i.e starts in one and ends in another.
 	indices := findIndices(spanIndexPrefix, startTime.Add(-time.Hour), endTime.Add(time.Hour))
 
-	results, err := s.client.MultiSearch().
-		Add(searchRequests...).
-		Index(indices...).
-		Do(s.ctx)
-	if err != nil {
-		return nil, err
-	}
+	var nextTime uint64
 
-	var traces []*model.Trace
-	for _, result := range results.Responses {
-		if result.Hits == nil || len(result.Hits.Hits) == 0 {
-			continue
+	searchAfterTime := make(map[string]uint64)
+	totalDocumentsFetched := make(map[string]int)
+	tracesMap := make(map[string]*model.Trace)
+	for {
+		if traceIDs == nil || len(traceIDs) == 0 {
+			break
 		}
-		spans, err := s.collectSpans(result.Hits.Hits)
+
+		for i, traceID := range traceIDs {
+			query := elastic.NewTermQuery("traceID", traceID)
+			if val, ok := searchAfterTime[traceID]; ok {
+				nextTime = val
+			}
+			searchRequests[i] = elastic.NewSearchRequest().IgnoreUnavailable(true).Type("span").Source(elastic.NewSearchSource().Query(query).Size(defaultDocCount).Sort("startTime", true).SearchAfter(nextTime))
+		}
+		// set traceIDs to empty
+		traceIDs = nil
+		results, err := s.client.MultiSearch().Add(searchRequests...).Index(indices...).Do(s.ctx)
```

Comment:
Is the 10000 limit for the entire query? ie if we search for 10 traces and the first trace takes up all 10000 spans, will this return a response for only the first trace?

Reply:
no, it's per trace. obviously, there is the downside of if there are 2 traces and 1st trace gets only 100 spans and next like 50K spans, entire query is going to wait to fetch 50,100 spans before it returns.